java-checks/src/main/java/org/sonar/java/checks/HardcodedURICheck.java

```diff
@@ -67,20 +77,83 @@ public class HardcodedURICheck extends IssuableSubscriptionVisitor {
   private static final Pattern URI_PATTERN = Pattern.compile(URI_REGEX + "|" + LOCAL_URI + "|" + DISK_URI + "|" + BACKSLASH_LOCAL_URI);
   private static final Pattern VARIABLE_NAME_PATTERN = Pattern.compile("filename|path", Pattern.CASE_INSENSITIVE);
   private static final Pattern PATH_DELIMETERS_PATTERN = Pattern.compile("\"/\"|\"//\"|\"\\\\\\\\\"|\"\\\\\\\\\\\\\\\\\"");
+  private static final Pattern RELATIVE_URI_PATTERN = Pattern.compile("^(/[\\w-+!*.]+){1,2}");
+
+
+  // we use these variables to track when we are visiting an annotation
+  private final Deque<AnnotationTree> annotationsStack = new ArrayDeque<>();
+
+  private record IdentifierData(Symbol symbol, String identifier) {
+  }
+
+  private final List<IdentifierData> identifiersUsedInAnnotations = new ArrayList<>();
+
+  private record VariableData(Symbol symbol, String identifier, ExpressionTree initializer) {
+  }
+
+  private final List<VariableData> hardCodedUri = new ArrayList<>();
+
+  @Override
+  public void setContext(JavaFileScannerContext context) {
+    super.setContext(context);
+    annotationsStack.clear();
+    identifiersUsedInAnnotations.clear();
+    hardCodedUri.clear();
+  }
+
+  @Override
+  public void leaveFile(JavaFileScannerContext context) {
+    // now, we know all variable that are used in annotation so we can report issues
+    Set<Symbol> idSymbols = identifiersUsedInAnnotations.stream()
+      .map(IdentifierData::symbol)
+      .filter(s -> !s.isUnknown())
+      .collect(Collectors.toSet());
+    Set<String> idNamesWithSemantic = identifiersUsedInAnnotations.stream()
+      .filter(i -> !i.symbol().isUnknown())
+      .map(IdentifierData::identifier)
+      .collect(Collectors.toSet());
+    Set<String> idNamesWithoutSemantic = identifiersUsedInAnnotations.stream()
+      .filter(i -> i.symbol().isUnknown())
+      .map(IdentifierData::identifier)
+      .collect(Collectors.toSet());
+
+    hardCodedUri.stream()
```

Comment:
Complex expressions are harder to read than simple if-then-else. I played in the IDE and came up with the following:
```
    for(VariableData v: hardCodedUri) {
      // equals to an identifier with unknown semantic, we cannot compare their symbols
      if(idNamesWithoutSemantic.contains(v.identifier())) {
        continue;
      }

      // idNamesWithSemantic is used to only compare the symbols when their string identifier are the same
      // as comparing symbols is costly
      if(idNamesWithSemantic.contains(v.identifier()) && idSymbols.contains(v.symbol())) {
        continue;
      }

      reportHardcodedURI(v.initializer());
    }
```
How about changing the code?

Reply:
:+1: